### ISSUE TITLE
Incorrect stage value for ERKs with c_s=1 but not FSAL

### ISSUE THREAD

**Issue by Steven-Roberts** [LLNL/sundials#311]
At the end of an ERK step, `erkStep_FullRHS` is called in `ARK_FULLRHS_END` mode to compute $f(y_{n+1})$. If the method has the first-same-as-last property (FSAL), there's no need to reevaluate $f$. The FSAL check

https://github.com/LLNL/sundials/blob/1ea097bb3bce207335ac35f0b5e78df5d71c6409/src/arkode/arkode_erkstep.c#L670-L672

is incorrect, however. It needs to check that the last row of $A$ is $b^T$. Several methods have $c_s=1$ but are not FSAL including `ARKODE_HEUN_EULER_2_1_2` and all the `ARKODE_ARK*` methods. A test with Heun's method confirms it's calling $f$ half as many times as it should. I am happy to make a PR fix, or should this be included in #293?

**Comment by Steven-Roberts** [comment]
Looks like we can revert to https://github.com/LLNL/sundials/commit/ed6a3a0609984dc4921c3affe5c4112578cb551b#diff-fcff12dbbb1374c54928dbaf2d6d89e890c283d7ad950c7f13b3cb1ef44910b3L653

**Comment by drreynolds** [comment]
I'm not sure that I understand what occurred.  Was this always broken (or at least for a long time), or was this introduced recently?  The reversion commit that you reference seems to be in relation to PR #293, but that hasn't yet been merged.

In my opinion, if this has been around for a while, then I think that @Steven-Roberts can put a fix into #293 then we can make it a priority to review/merge this promptly.  If this bug was just introduced, however, then a bugfix release would be more appropriate.

**Comment by balos1** [comment]
@drreynolds The reversion commit Steven linked to is not linked to PR #293. It is linked to PR that was merged in BitBucket prior to our move to GitHub. The branch name was `arkode-fullrhs-enum`, a very similar name to PR #293.

**Comment by balos1** [comment]
Here is a link to the PR that made the change to just check `c[s-1]`.

**Comment by balos1** [comment]
I do not understand why yet, but this seems to only be a problem in ERKStep, and not ARKStep. With ARKStep and `ARKODE_HEUN_EULER_2_1_2` it uses the expected number of function calls.  

[Here](https://github.com/LLNL/sundials/blob/1ea097bb3bce207335ac35f0b5e78df5d71c6409/src/arkode/arkode_arkstep.c#L1429C1-L1434C30) is the check in ARKStep. 

```c
    /* determine if explicit/implicit RHS functions need to be recomputed */
    recomputeRHS = SUNFALSE;
    if ( step_mem->explicit && (SUNRabs(step_mem->Be->c[step_mem->stages-1]-ONE)>TINY) )
      recomputeRHS = SUNTRUE;
    if ( step_mem->implicit && (SUNRabs(step_mem->Bi->c[step_mem->stages-1]-ONE)>TINY) )
      recomputeRHS = SUNTRUE;
```

**Comment by balos1** [comment]
OK, so the reason ARKStep still makes the expected number of function evaluations is because it always starts the internal stage loop at the first stage, while ERKStep starts at the second stage. 

As noted in the code, the stage loop in ERKStep is written this way since it is assuming that a call to the RHS was made "at the start of the step", which should be the same as the first stage RHS. In reality, the call "at the start of the step" actually happens at the end of the previous step via a call to the fullRHS with the `ARK_FULLRHS_END` case.  The erroneous FSAL check results in the RHS not being called in `ARK_FULLRHS_END`, so ERKStep assumption is then, erroneously, violated. 

With ARKStep, since the loop starts with the first stage and the assumption that RHS is the same as "the start of the step", the erroneous FSAL check does not impact it. **However**, the correct FSAL check would actually result in **extra** calls to the RHS by ARKStep.

**Comment by Steven-Roberts** [comment]
> Was this always broken (or at least for a long time), or was this introduced recently?

The diff I linked suggests this was introduced about 2 years ago.

> if this has been around for a while, then I think that @Steven-Roberts can put a fix into https://github.com/LLNL/sundials/pull/293 then we can make it a priority to review/merge this promptly

Ok, I'll add a commit to #293 then

I had not checked ARKStep yet, so that's some helpful results, Cody. Although the loop start fixes the issue (at the cost of an extra function evaluation), I see a potential problem. For a non-FSAL ARK method with $c^E = c^I = 1$, the Hermite interpolant will use $f(Y_s)$. For most methods, this is only an order 1 approximation to $f(y(t_{n+1}))$, while $f(y_{n+1})$ is order $q$. I guess this is ok for low order interpolants, but for high order it's problematic. The fix would be similar to the fix for ERKStep.

**Comment by balos1** [comment]
@Steven-Roberts I am not suggesting that adjusting the stage loop start is a fix for the problem, rather that is how the code is right now.

**Comment by Steven-Roberts** [comment]
Yes, I understand. The ARKStep FASL issue I mention is present in the current code. I agree we'll need to think carefully about removing the extra function eval while fixing these bugs.

**Comment by balos1** [comment]
I started a [bugfix branch](https://github.com/LLNL/sundials/tree/bugfix/erkstep-arkstep-recomputerhs). Right now, it just has the changes I made to product the test result I showed above.

**Comment by gardner48** [comment]
Fixed in SUNDIALS 6.7.0
